nucliadb/nucliadb/ingest/orm/processor.py

```diff
@@ -372,64 +371,38 @@ async def txn(
-
-            if resource:
-                await resource.compute_global_text()
-                await resource.compute_global_tags(resource.indexer)
-                if message.reindex:
-                    # when reindexing, let's just generate full new index message
-                    resource.replace_indexer(await resource.generate_index_message())
-
-            if resource and resource.modified:
-                shard_id = await kb.get_resource_shard_id(uuid)
-                node_klass = get_node_klass()
-
-                if shard_id is not None:
-                    shard = await kb.get_resource_shard(shard_id, node_klass)
-
-                if shard is None:
-                    # It's a new resource, get current active shard to place
-                    # new resource on
-                    shard = await node_klass.get_current_active_shard(txn, kbid)
-                    if shard is None:
-                        # no shard available, create a new one
-                        similarity = await kb.get_similarity()
-                        shard = await node_klass.create_shard_by_kbid(
-                            txn, kbid, similarity=similarity
-                        )
-                    await kb.set_resource_shard_id(uuid, shard.sharduuid)
-
-                if shard is not None:
-                    counter = await shard.add_resource(
-                        resource.indexer.brain, seqid, partition=partition, kb=kbid
-                    )
-                    if (
-                        counter is not None
-                        and counter.fields > settings.max_shard_fields
-                    ):
-                        # shard is full, create a new one so next resource
-                        # is placed on a new shard
-                        similarity = await kb.get_similarity()
-                        shard = await node_klass.create_shard_by_kbid(
-                            txn, kbid, similarity=similarity
-                        )
-                else:
-                    raise AttributeError("Shard is not available")
+            if resource is None:
+                raise NoResourceError()
+
+            # TODO: in case of reindex = True, shouldn't we do that after generate_index_message?
+            await resource.compute_global_text()
+            await resource.compute_global_tags(resource.indexer)
+            if reindex:
+                # When reindexing, let's just generate full new index message
+                resource.replace_indexer(await resource.generate_index_message())
+
+            if resource.modified or reindex:
+                # TODO: Shouldn't we index the resource after it has been commited?
+                counter, shard = await self.index_resource(
+                    resource, txn, kb, uuid, seqid, partition
+                )
 
                 await txn.commit(partition, seqid)
 
+                # TODO: Look if it's really needed to commit the slug every time
```

Comment:
And what happens if slug commit fails but the other succeeded? This is weird here.